### ISSUE TITLE
Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by MuhammadSawalhy** [marionebl/svg-term-cli#71]
### Reproduce

```bash
svg-term --cast 47HiIPl0HQxrJ0R0vR9qpxzS4 --window --out animating-my-name.svg
```

### Description

After running this command, my computer was noticeably slow, my mouse was not reponding sometimes. I quickly tried to run `htop` command in Linux:

![Screenshot from 2021-03-24 22-34-22](https://user-images.githubusercontent.com/42011920/112380352-f88e1480-8cf1-11eb-86f7-a9fc96072e3f.png)

After waiting a while, Nodejs logged these lines:

```
<--- Last few GCs --->

[313356:0x513b270]    89040 ms: Scavenge 1892.2 (2064.9) -> 1879.9 (2064.9) MB, 6.7 / 0.0 ms  (average mu = 0.255, current mu = 0.145) allocation failure 
[313356:0x513b270]    89205 ms: Scavenge 1896.4 (2064.9) -> 1886.5 (2067.6) MB, 24.6 / 0.0 ms  (average mu = 0.255, current mu = 0.145) allocation failure 
[313356:0x513b270]    90199 ms: Mark-sweep 1904.1 (2067.6) -> 1880.2 (2068.9) MB, 878.9 / 0.0 ms  (average mu = 0.299, current mu = 0.342) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa04200 node::Abort() [svg-term]
 2: 0x94e4e9 node::FatalError(char const*, char const*) [svg-term]
 3: 0xb7978e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [svg-term]
 4: 0xb79b07 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [svg-term]
 5: 0xd34395  [svg-term]
 6: 0xd34f1f  [svg-term]
 7: 0xd42fab v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [svg-term]
 8: 0xd46b6c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [svg-term]
 9: 0xd1524b v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [svg-term]
10: 0x105b23f v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [svg-term]
11: 0x1401219  [svg-term]
```

**Comment by christos-h** [comment]
@MuhammadSawalhy  Just ran into this. It turns out the Node runtime is running out of memory (I think the default is 500mb or something). Depending on how much memory you have on your machine, you can increase this by setting environment variables. For a 4GB heap you can do this for example:

```bash
NODE_OPTIONS=--max-old-space-size=4096 svg-term --cast=236096 --out examples/parrot.svg --window
```

I can see that you were already close to the limit of your mem utilisation. The next time you run a conversion you may want to close all other windows and increase the size of your swap.

Good luck.

**Comment by cjbrigato** [comment]
Having memory issues doesn't always mean "lacking memory".
A good starting point to **actually** solve these issues is to understand that what lacks here is more about wishes to optimize and deliver quality software than memory.

> "The next time you run a conversion you may want to close all other windows and increase the size of your swap."

I'll try that on our headless server farm, thanks.

**Comment by assawalhy** [comment]
> **wishes to optimize and deliver quality software than memory**

This is the purpose of the issue 😅

**Comment by cjbrigato** [comment]
Indeed, but as expected it was not understood in proper way. Hence my addition ;)

**Comment by christos-h** [comment]
> A good starting point to actually solve these issues is to understand that what lacks here is more about wishes to optimize and deliver quality software than memory.

Of course you're right. I was merely trying to offer pointers to individuals who arrived here by googling an error, and just want to get the job done without necessarily having the time to address the underlying issues.

**Comment by cjbrigato** [comment]
We tried converting a 1:30 asciinema with SVG term giving more than 4gb of ram available. It just killed the machine up to no responsiveness and needed reset.
In the meanwhile, https://github.com/nbedos/termtosvg just rendered the asciicast in under 5 sec flawlessly with under 200mb of ram.

The difference is so huge and the answer was so out of the scope that I first thought your answer was a joke. Definitely thought it was when I read the last sentence.

If it was not, then sorry, I never meant to talk about optimization seriously on a node-js software.

**Comment by helmut-messerer** [comment]
see
https://github.com/socketio/socket.io/issues/3477#issuecomment-677801998

**Comment by helmut-messerer** [comment]
see https://github.com/socketio/socket.io/issues/3477#issuecomment-677801998
try install libjemalloc and use
LD_PRELOAD=/path/to/libjemalloc.so
before launching javascript with node
